src/sentry/api/endpoints/project_dynamic_sampling.py

```diff
@@ -281,48 +347,128 @@ def get(self, request: Request, project) -> Response:
-                query=f"event.type:transaction trace:[{','.join(trace_id_list)}]",
-                params={
-                    "start": query_time_range.start_time,
-                    "end": query_time_range.end_time,
-                    "project_id": list(projects_in_org),
-                    "organization_id": project.organization.id,
-                },
-                limit=20,
-                referrer="dynamic-sampling.distribution.fetch-project-breakdown",
-            )
-
+            # Discover query to fetch parent projects (smart query)
+            projects_counts = self.__project_stats_query(project, query_time_range, trace_ids)
             # If the number of the projects in the breakdown is greater than 10 projects,
             # then a question needs to be raised on the eligibility of the org for dynamic sampling
-            if len(project_breakdown) > 10:
+            if len(projects_counts) > 10:
                 return Response(
                     status=status.HTTP_400_BAD_REQUEST,
                     data={
                         "detail": "Way too many projects in the distributed trace's project breakdown"
                     },
                 )
 
+            # Requesting for a project that is root for some transactions,
+            # but not root for others and part of distributed trace like
+            root_projects_count = len(
+                [project for project in projects_counts if project["root"] > 0]
+            )
+            # get requested project id in projects_count dataset
+            requested_project = next(
+                item for item in projects_counts if item["project_id"] == project.id
+            )
+
+            # This condition covers all 3 cases (A, C, D):
+            if root_projects_count == 1:
+                # build parent_project_breakdown
+                parent_project_breakdown = [
+                    {"project_id": project.id, "project": project.slug, "percentage": 1}
+                ]
+            # Case B: the most tricky one and requires extra query
+            elif root_projects_count > 1 and requested_project["root"] > 0:
+                parent_project_breakdown = []
+                total_count = sum(
+                    _project["count"] for _project in projects_counts if _project["root"] > 0
+                )
+                for _project in projects_counts:
+                    if _project["root"] > 0:
+                        parent_project_breakdown.append(
+                            {
+                                "project": _project["project"],
+                                "project_id": _project["project_id"],
+                                "percentage": _project["root"] / total_count
+                                if total_count > 0
+                                else 0,
+                            }
+                        )
+
+            else:
+                parent_project_breakdown = []
```

Comment:
I think this logic can be simplified into this
```suggestion
            if root_projects_count > 0:
                parent_project_breakdown = []
                total_count = sum(
                    _project["count"] for _project in projects_counts if _project["root"] > 0
                )
                for _project in projects_counts:
                    if _project["root"] > 0:
                        parent_project_breakdown.append(
                            {
                                "project": _project["project"],
                                "project_id": _project["project_id"],
                                "percentage": _project["root"] / total_count
                                if total_count > 0
                                else 0,
                            }
                        )

            else:
                parent_project_breakdown = []
```